navigation-fragments/src/main/java/dev/hotwire/navigation/util/HotwireViewScreenshotHolder.kt

```diff
@@ -0,0 +1,115 @@
+import kotlinx.coroutines.suspendCancellableCoroutine
+
+internal class HotwireViewScreenshotHolder {
+    private var bitmap: Bitmap? = null
+    private var screenshotOrientation = 0
+    private var screenshotZoomed = false
+    var currentlyZoomed = false
+
+    fun reset() {
+        bitmap = null
+        screenshotOrientation = 0
+        screenshotZoomed = false
+    }
+
+    fun showScreenshotIfAvailable(hotwireView: HotwireView) {
+        if (screenshotOrientation == hotwireView.currentOrientation() &&
+            screenshotZoomed == currentlyZoomed
+        ) {
+            bitmap?.let { hotwireView.addScreenshot(it) }
+        }
+    }
+
+    suspend fun captureScreenshot(hotwireView: HotwireView) {
+        bitmap = copyViewToBitmap(hotwireView)
+        screenshotOrientation = hotwireView.currentOrientation()
+        screenshotZoomed = currentlyZoomed
+    }
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    private suspend fun copyViewToBitmap(hotwireView: HotwireView): Bitmap? {
+        return suspendCancellableCoroutine { continuation ->
+            val start = System.currentTimeMillis()
+            val window = hotwireView.getActivity()?.window
+
+            if (window == null || !hotwireView.isLaidOut || !hasEnoughMemoryForScreenshot() ||
+                hotwireView.width <= 0 || hotwireView.height <= 0
+            ) {
+                if (continuation.isActive) {
+                    continuation.resume(null, null)
+                }
+                return@suspendCancellableCoroutine
+            }
+
+            val rect = Rect()
+            hotwireView.getGlobalVisibleRect(rect)
+
+            val bitmap = Bitmap.createBitmap(rect.width(), rect.height(), Bitmap.Config.ARGB_8888)
+
+            try {
+                PixelCopy.request(
+                    window, rect, bitmap,
+                    { result ->
+                        if (result == PixelCopy.SUCCESS) {
+                            logEvent(
+                                "viewScreenshotCreated", listOf(
+                                    "size" to "${bitmap.width}x${bitmap.height}",
+                                    "duration" to "${System.currentTimeMillis() - start}ms",
+                                )
+                            )
+                            if (continuation.isActive) {
+                                continuation.resume(bitmap, null)
+                            }
+                        } else {
+                            logEvent("viewScreenshotFailed", listOf("error" to result))
+                            if (continuation.isActive) {
+                                continuation.resume(null, null)
+                            }
+                        }
+                    },
+                    Handler(Looper.getMainLooper())
+                )
+            } catch (exception: Exception) {
+                logEvent("viewScreenshotFailed", listOf("error" to exception))
```

Comment:
Note that there's a `logError()` function available that handles logging exceptions.